src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -554,12 +669,131 @@ private static void TransformFeatures(IExceptionContext ectx, ref VBuffer<Float>
+            var view = PcaTransform.Create(h, input, input.Data);
             return new CommonOutputs.TransformOutput()
             {
                 Model = new TransformModel(h, view, input.Data),
                 OutputData = view
             };
         }
     }
+
+    public sealed class PcaEstimator : IEstimator<PcaTransform>
+    {
+        internal static class Defaults
+        {
+            public const string WeightColumn = null;
+            public const int Rank = 20;
+            public const int Oversampling = 20;
+            public const bool Center = true;
+            public const int Seed = 0;
+        }
+
+        private readonly IHost _host;
+        private readonly PcaTransform.ColumnInfo[] _columns;
+
+        /// <summary>Convinence constructor for simple one column case.</summary>
+        /// <include file='doc.xml' path='doc/members/member[@name="PCA"]/*'/>
+        /// <param name="env">The environment.</param>
+        /// <param name="inputColumn">Input column to apply PCA on.</param>
+        /// <param name="outputColumn">Output column. Null means <paramref name="inputColumn"/> is replaced.</param>
+        /// <param name="weightColumn">The name of the weight column.</param>
+        /// <param name="rank">The number of components in the PCA.</param>
+        /// <param name="overSampling">Oversampling parameter for randomized PCA training.</param>
+        /// <param name="center">If enabled, data is centered to be zero mean.</param>
+        /// <param name="seed">The seed for random number generation</param>
+        public PcaEstimator(IHostEnvironment env, string inputColumn, string outputColumn = null,
+            string weightColumn = Defaults.WeightColumn, int rank = Defaults.Rank,
+            int overSampling = Defaults.Oversampling, bool center = Defaults.Center,
+            int? seed = null)
+            : this(env, new PcaTransform.ColumnInfo(inputColumn, outputColumn ?? inputColumn, weightColumn, rank, overSampling, center, seed))
+        {
+        }
+
+        public PcaEstimator(IHostEnvironment env, params PcaTransform.ColumnInfo[] columns)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(nameof(PcaEstimator));
+            _columns = columns;
+        }
+
+        public PcaTransform Fit(IDataView input) => new PcaTransform(_host, input, _columns);
+
+        public SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+            var result = inputSchema.Columns.ToDictionary(x => x.Name);
+            foreach (var colInfo in _columns)
+            {
+                if (!inputSchema.TryFindColumn(colInfo.Input, out var col))
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.Input);
+
+                if (col.Kind != SchemaShape.Column.VectorKind.Vector || col.ItemType.RawKind != DataKind.R4)
```

Comment:
>ItemType [](start = 76, length = 8)

`!col.ItemType.Equals(NumberType.R4)` is a more direct comparison to `R4`

Reply:
done

---
In reply to: [227593906](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227593906) [](ancestors = 227593906)